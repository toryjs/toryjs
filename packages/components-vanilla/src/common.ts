export type ReactComponent = string | React.ComponentType;
