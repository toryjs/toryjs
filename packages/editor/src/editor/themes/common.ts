export { Theme } from '@toryjs/ui';
