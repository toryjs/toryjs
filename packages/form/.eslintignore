node_modules
.fusebox
.cache
luis
public
dist
summary.json
snapshots.js
summary.ts
coverage