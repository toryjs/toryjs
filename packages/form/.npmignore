node_modules
src
wallaby.js
yarn-error.log
.eslintignore
.eslintrc.js
jest.config.js
.vscode
.github
jest.config.js
mocha.config.js