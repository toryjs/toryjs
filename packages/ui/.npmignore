luis
src
node_modules
.gitignore
index.html
luis.fuse.js
luis.fuse.html
luis.parcel.html
mocha.config.js
upgrade.js
wallaby.js
yarn-error.log
.eslintignore
.eslintrc.js
jest.config.js
public

