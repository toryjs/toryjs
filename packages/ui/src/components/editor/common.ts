import React from 'react';

export function calculatePosition(layout: string, e: React.DragEvent) {
  return layout === 'row' ? calculateVerticalPosition(e) : calculateHorizontalPosition(e);
}

export function calculateHorizontalPosition(e: React.DragEvent) {
  let element = e.currentTarget;
  let elementLeft = window.scrollX + element.getBoundingClientRect().left;
  let elementWidth = element.clientWidth;

  return e.pageX - elementLeft < elementWidth / 2 ? 'left' : 'right';
}

export function calculateVerticalPosition(e: React.DragEvent) {
  let element = e.currentTarget;
  let elementTop = window.scrollX + element.getBoundingClientRect().top;
  let elementHeight = element.clientHeight;
  let dragStart = e.pageY;

  // console.log(`${id} ${dragStart} - ${elementTop} < ${elementHeight} | ${dragStart - elementTop}`);

  return dragStart - elementTop < elementHeight / 2 ? 'top' : 'bottom';
}
