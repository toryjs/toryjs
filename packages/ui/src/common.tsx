import React from 'react';
import names from 'classnames';

import { interpolate } from '@tomino/toolbelt';
import { DataSet, FormElement, FormComponentProps, FormComponent } from '@toryjs/form';
import { BoundProp, BoundType } from '@toryjs/form';

import { css } from 'emotion';
import { toJS as mobxToJs } from 'mobx';
export { default as names } from 'classnames';

import { getPropValue } from './helpers';
import { ContextType } from './context';

export { Theme } from './editor/themes/common';
export { Context } from './context';
export {
  handle,
  getValue,
  isNullOrEmpty,
  prop,
  processControl,
  valueSource,
  createComponents
} from './helpers';
export { DynamicComponent as DynamicControl, paintProps } from './components/dynamic_component';

export const select = css`
  .react-select__control {
    border: 0px;
    border-radius: 0 !important;
    min-height: 23px;
  }

  .react-select__indicators {
    height: 23px;
  }

  .react-select__indicator {
    padding: 2px;
  }

  .react-select__value-container {
    height: 23px;

    > div {
      /* position: absolute; */
      margin: 0px;
      padding: 0px;
    }
  }

  .react-select__single-value {
    margin-top: 0px;
  }

  .react-select__menu {
    border-radius: 0px;
    margin-top: 2px;
  }

  label: select;
`;

export const fieldSet = css`
  /* name: fieldset */
  border: solid 1px #ddd;
  border-radius: 5px;
  background-color: #fafafa;

  legend {
    padding: 0px 6px;
    font-weight: bold;
  }
`;

export const breakingLabel = css`
  display: block;
  font-weight: bold;
  margin-top: 3px;
  margin-bottom: 3px;
`;

export const nonBreakingLabel = css`
  display: inline-block;
  font-weight: bold;
  padding-right: 6px;
  margin-top: 3px;
  margin-bottom: 3px;
`;

export const fullWidth = css`
  width: 100%;
`;

export const noPadding = css`
  padding: 0px !important;
`;

export const marginLess = css`
  margin: 0px !important;
`;

export const pointer = css`
  cursor: pointer !important;
  label: pointer;
`;

export const error = css`
  background: red;
  padding: 3px;
  border-radius: 3px;
  font-weight: bold;
  color: white;
`;

export type SignatureType = {
  comment: string;
  signature: string;
  verifiedState: 'Pending' | 'Verified' | 'Rejected';
  rejected?: boolean;
  uid: string;
  name: string;
  date: Date;
  setValue?(name: string, value: any): any;
};

export type SignatureHandlers = {
  validateForm(): boolean;
  verifySignature(owner: any, args: { uid: string; signature: string }): Promise<string>;
  sign(
    owner: DataSet,
    args: {
      source: string;
      reject: boolean;
      password: string;
      reason: string;
      submit: boolean;
    }
  ): Promise<SignatureType>;
  signatureFont(): string;
};

export type FormEditorProps = {
  formElement: DynamicForm.FormDataSet;
  formData: DataSet;
  handlers?: { [index: string]: any };
  readOnly: boolean;
};

export function parseFromOwner(str: string, owner: DataSet) {
  try {
    return interpolate(str, owner);
  } catch (ex) {
    return 'Error parsing string: ' + ex.message;
  }
}

export function sourceValue(prop: BoundProp) {
  return prop ? (prop as BoundType).source : null;
}

export function handlerValue(prop: BoundProp) {
  return prop ? (prop as BoundType).handler : null;
}

// let onClick: (props: FormComponentProps) => Function;
// export function registerOnClick(onClickHandler: typeof onClick) {
//   onClick = onClickHandler;
// }

export function root(formElement: FormElement) {
  let root = formElement;
  while (root.parent != null && (!root.pages || root.pages.length === 0)) {
    root = root.parent;
  }
  return root;
}

const extraProps = ['editorLabel'];

export function cleanProps(props: any) {
  const result: any = {};
  const js = toJS(props);
  for (let key of Object.keys(js)) {
    if (js[key] === '' || js[key] == null || js[key].length === 0 || extraProps.indexOf(key) >= 0) {
      continue;
    }
    result[key] = js[key];
  }
  return result;
}

export function tryInterpolate(text: string, owner: DataSet) {
  try {
    return interpolate(text, owner);
  } catch (ex) {
    return 'Error: ' + ex.message;
  }
}

export function useMergeState<T>(initialState: T): [T, (t: Partial<T>) => void] {
  const [state, setState] = React.useState(initialState);
  const setMergedState = (newState: any) =>
    setState((prevState: any) => Object.assign({}, prevState, newState));
  return [state, setMergedState];
}

// export function createComponents(props: FormComponentProps, className: string = null) {
//   if (!props.formElement.elements || props.formElement.elements.length === 0) {
//     return undefined;
//   }
//   return props.formElement.elements.map((e, i) => (
//     <React.Fragment key={i}>
//       {props.catalogue.createComponent(
//         { ...props, className: '' },
//         e,
//         className
//         // extraProps
//         // null,
//         // onClick ? onClick(props) : undefined
//       )}
//     </React.Fragment>
//   ));
// }

// type Options = {
//   drawLabel?: boolean;
//   className?: string;
//   onClick?: any
// }

export function createComponent1(
  props: FormComponentProps,
  formElement: FormElement,
  context: ContextType,
  className?: string,
  extraProps?: any
) {
  if (!formElement || formElement.control === 'EditorCell') {
    return null;
  }

  if (props.catalogue.components[formElement.control] === undefined) {
    return <div className={names(className, error)}>{formElement.control}</div>;
  }

  const hide = getPropValue(props, formElement, context, 'hidden', false);
  if (hide) {
    return null;
  }

  const controlDefinition = props.catalogue.components[formElement.control];
  const Control =
    (controlDefinition as FormComponent).Component ||
    (controlDefinition as React.ComponentType<any>);

  Control.displayName = 'CreateComponentWrapper';

  // TODO: Make this type safe by creating a render method in catalogue
  let renderedControl = (
    <Control
      owner={props.owner}
      className={className}
      handlers={props.handlers}
      catalogue={props.catalogue}
      formElement={formElement}
      extra={{ ...props.extra, ...extraProps }}
      dataProps={props.dataProps}
    />
  );

  return renderedControl;
}

export function toJS<T>(obj: T): T {
  return mobxToJs(obj);
}

export function addProviders(mainForm: FormElement, form: FormElement) {
  let formElement = form;

  // add all root providers from the man form
  if (form !== mainForm) {
    let model = toJS(mainForm);
    let provider = model;

    while (
      provider.elements &&
      provider.elements.length &&
      provider.elements[0].control.indexOf('Provider') >= 0
    ) {
      provider = provider.elements[0];
    }

    if (provider != model) {
      provider.elements = [form];
      formElement = model;
    }
  }

  return formElement;
}

// export function props<T>(props: T, defaultProps?: Partial<T>): T {
//   return controlProps || (defaultProps as any) || {};
// }
