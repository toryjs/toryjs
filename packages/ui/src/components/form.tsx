import React from 'react';

import { FormModel, FormComponentCatalogue, Handlers } from '@toryjs/form';
import { IProject, IStorage } from '../storage/common_storage';
import { FormView } from './form_view';
import { Context, context } from '../context';
import { formDatasetToJS, schemaDatasetToJS } from '../helpers';

type Props = {
  project?: IProject;
  storage?: IStorage;
  catalogue: FormComponentCatalogue;
  handlers: Handlers<any, any>;
};

export const ToryForm: React.FC<Props> = props => {
  const [project, setProject] = React.useState<IProject>(props.project ? props.project : null);
  const formModel = project
    ? new FormModel(formDatasetToJS(project.form), schemaDatasetToJS(project.schema, false), {})
    : null;

  if (!project) {
    if (props.storage) {
      props.storage.loadProject().then(p => setProject(p));
    } else {
      throw new Error('You need to provide either a storage or a form model');
    }
    return <div>Loading ...</div>;
  }
  return (
    <Context.Provider value={context}>
      <FormView
        formElement={formModel.formDefinition}
        owner={formModel.dataSet}
        catalogue={props.catalogue}
        handlers={props.handlers}
      />
    </Context.Provider>
  );
};

ToryForm.displayName = 'ToryForm';
