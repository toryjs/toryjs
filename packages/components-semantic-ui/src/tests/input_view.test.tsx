import { JSONSchema } from '@toryjs/form';

import { create } from '@toryjs/ui';
import { prepareComponent } from './common';

const schema: JSONSchema = {
  type: 'object',
  properties: {
    name: {
      type: 'string'
    },
    age: {
      type: 'number',
      minimum: 30
    },
    younger: {
      type: 'string',
      expression: `console.log(this); return this['age'] - 20`
    },
    older: {
      type: 'string',
      expression: `this['age'] + 10`
    }
  },
  required: ['name']
};

const formDefinition = create.grid({
  props: {
    columns: 4,
    rows: 2
  },
  elements: [
    {
      control: 'Input',
      props: {
        row: 1,
        column: 1,
        width: 2,
        inputLabel: 'Name',
        label: 'Mimo',
        value: { source: 'name' }
      }
    },
    {
      props: { row: 2, column: 1, width: 1, value: { source: 'age' }, label: 'Age: ' },
      control: 'Input'
    },
    {
      props: { row: 2, column: 3, width: 2, value: { source: 'younger' }, label: 'Younger:' },
      control: 'Input'
    }
    // {
    //   props: {
    //     row: 1,
    //     column: 3,
    //     width: 2,
    //     value: { source: 'older' },
    //     label: 'Older',
    //     inline: true
    //   },
    //   control: 'Input'
    // }
  ]
});

const formData = {
  name: 'Tomas',
  age: 33
};

describe('Input', () => {
  // it('renders tab menu', () => {
  //   const wrapper = renderer.create(component());
  //   expect(wrapper).toMatchSnapshot();
  // });

  return {
    component: prepareComponent(formDefinition, schema, formData, {
      addEditor: true,
      showStandard: true
    })
  };
});
