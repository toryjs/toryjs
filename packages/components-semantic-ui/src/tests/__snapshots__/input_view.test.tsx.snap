exports["Input changes value and all related formulas 1"] = "\n<div\n  className=\"ui segment ui form\"\n>\n  <div\n    className=\"ui dividing header\"\n  >\n    Standard\n  </div>\n  <div\n    className=\"ui form \"\n  >\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"fourteen wide field\"\n      >\n        <label\n          htmlFor=\"name\"\n        >\n          Mimo\n        </label>\n        <div\n          className=\"ui labeled input\"\n        >\n          <div\n            className=\"ui label label\"\n            onClick={[Function]}\n          >\n            Name\n          </div>\n          <input\n            name=\"name\"\n            onChange={[Function]}\n            type=\"text\"\n            value=\"Tomas\"\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n        <label\n          htmlFor=\"older\"\n        >\n          Older\n        </label>\n        <div\n          className=\"ui input\"\n        >\n          <input\n            name=\"older\"\n            onChange={[Function]}\n            type=\"text\"\n            value={50}\n          />\n        </div>\n      </div>\n    </div>\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"one wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"age\"\n        >\n          Age: \n        </label>\n        <div\n          className=\"ui input\"\n        >\n          <input\n            name=\"age\"\n            onChange={[Function]}\n            type=\"text\"\n            value={40}\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"younger\"\n        >\n          Younger:\n        </label>\n        <div\n          className=\"ui input\"\n        >\n          <input\n            name=\"younger\"\n            onChange={[Function]}\n            type=\"text\"\n            value={20}\n          />\n        </div>\n      </div>\n    </div>\n  </div>\n  <div\n    className=\"ui dividing header\"\n  >\n    Readonly\n  </div>\n  <div\n    className=\"ui form form-Abfdcefijjc\"\n  >\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"fourteen wide field\"\n      >\n        <label\n          htmlFor=\"name\"\n        >\n          Mimo\n        </label>\n        <div\n          className=\"ui disabled labeled input\"\n        >\n          <div\n            className=\"ui label label\"\n            onClick={[Function]}\n          >\n            Name\n          </div>\n          <input\n            disabled={true}\n            name=\"name\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value=\"Tomas\"\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n        <label\n          htmlFor=\"older\"\n        >\n          Older\n        </label>\n        <div\n          className=\"ui disabled input\"\n        >\n          <input\n            disabled={true}\n            name=\"older\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value={50}\n          />\n        </div>\n      </div>\n    </div>\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"one wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"age\"\n        >\n          Age: \n        </label>\n        <div\n          className=\"ui disabled input\"\n        >\n          <input\n            disabled={true}\n            name=\"age\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value={40}\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"younger\"\n        >\n          Younger:\n        </label>\n        <div\n          className=\"ui disabled input\"\n        >\n          <input\n            disabled={true}\n            name=\"younger\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value={20}\n          />\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n";

exports["Input renders correctly 1"] = "\n<div\n  className=\"ui segment ui form\"\n>\n  <div\n    className=\"ui dividing header\"\n  >\n    Standard\n  </div>\n  <div\n    className=\"ui form \"\n  >\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"fourteen wide field\"\n      >\n        <label\n          htmlFor=\"name\"\n        >\n          Mimo\n        </label>\n        <div\n          className=\"ui labeled input\"\n        >\n          <div\n            className=\"ui label label\"\n            onClick={[Function]}\n          >\n            Name\n          </div>\n          <input\n            name=\"name\"\n            onChange={[Function]}\n            type=\"text\"\n            value=\"Tomas\"\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n        <label\n          htmlFor=\"older\"\n        >\n          Older\n        </label>\n        <div\n          className=\"ui input\"\n        >\n          <input\n            name=\"older\"\n            onChange={[Function]}\n            type=\"text\"\n            value={43}\n          />\n        </div>\n      </div>\n    </div>\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"one wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"age\"\n        >\n          Age: \n        </label>\n        <div\n          className=\"ui input\"\n        >\n          <input\n            name=\"age\"\n            onChange={[Function]}\n            type=\"text\"\n            value={33}\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"younger\"\n        >\n          Younger:\n        </label>\n        <div\n          className=\"ui input\"\n        >\n          <input\n            name=\"younger\"\n            onChange={[Function]}\n            type=\"text\"\n            value={13}\n          />\n        </div>\n      </div>\n    </div>\n  </div>\n  <div\n    className=\"ui dividing header\"\n  >\n    Readonly\n  </div>\n  <div\n    className=\"ui form form-Abfdcefijjc\"\n  >\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"fourteen wide field\"\n      >\n        <label\n          htmlFor=\"name\"\n        >\n          Mimo\n        </label>\n        <div\n          className=\"ui disabled labeled input\"\n        >\n          <div\n            className=\"ui label label\"\n            onClick={[Function]}\n          >\n            Name\n          </div>\n          <input\n            disabled={true}\n            name=\"name\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value=\"Tomas\"\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n        <label\n          htmlFor=\"older\"\n        >\n          Older\n        </label>\n        <div\n          className=\"ui disabled input\"\n        >\n          <input\n            disabled={true}\n            name=\"older\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value={43}\n          />\n        </div>\n      </div>\n    </div>\n    <div\n      className=\"fields\"\n    >\n      <div\n        className=\"one wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"age\"\n        >\n          Age: \n        </label>\n        <div\n          className=\"ui disabled input\"\n        >\n          <input\n            disabled={true}\n            name=\"age\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value={33}\n          />\n        </div>\n      </div>\n      <div\n        className=\"two wide field\"\n      >\n         \n      </div>\n      <div\n        className=\"inline field\"\n      >\n        <label\n          htmlFor=\"younger\"\n        >\n          Younger:\n        </label>\n        <div\n          className=\"ui disabled input\"\n        >\n          <input\n            disabled={true}\n            name=\"younger\"\n            onChange={[Function]}\n            tabIndex={-1}\n            type=\"text\"\n            value={13}\n          />\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n";

